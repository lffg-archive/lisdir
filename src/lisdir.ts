export * from './list-directory';
export * from './list-modules';
