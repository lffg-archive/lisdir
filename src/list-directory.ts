import fs, { Dirent } from 'fs';
import { join } from 'path';
import { promisify } from 'util';

const readdir = promisify(fs.readdir);

export interface Entry extends Dirent {
  path: string;
}

export interface ListDirectoryOptions {
  recursive: boolean;
  filter: (file: Entry) => boolean;
}

export async function listDirectory(
  dir: string,
  options?: Partial<ListDirectoryOptions>
) {
  const config: ListDirectoryOptions = {
    // Default options:
    recursive: false,
    filter: () => true,

    // User options:
    ...options
  };

  const list: Entry[] = [];

  const dirents = await readdir(dir, { withFileTypes: true });

  for (const dirent of dirents) {
    const entry: Entry = Object.assign(dirent, {
      path: join(dir, dirent.name)
    });

    if (entry.isDirectory() && config.recursive) {
      list.push(...(await listDirectory(entry.path, config)));
    } else {
      if (config.filter(entry)) {
        list.push(entry);
      }
    }
  }

  return list;
}
