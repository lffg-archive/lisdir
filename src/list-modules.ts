import { extname } from 'path';
import { ListDirectoryOptions, listDirectory } from './list-directory';

export interface ModuleEntry<T = any> {
  name: string;
  path: string;
  module: T;
}

export interface ListModulesOptions extends ListDirectoryOptions {
  extensions: string[];
}

export async function listModules<T = any>(
  dir: string,
  options?: Partial<ListModulesOptions>
): Promise<ModuleEntry<T>[]> {
  const config: ListModulesOptions = {
    // Default options:
    recursive: false,
    filter: () => true,
    extensions: ['js', 'jsx', 'ts', 'tsx'],

    // User options:
    ...options
  };

  const extensions = config.extensions.map((ext) =>
    /^\./.test(ext) ? ext : `.${ext}`
  );

  const files = await listDirectory(dir, {
    recursive: config.recursive,
    filter: (file) =>
      file.isFile() &&
      extensions.includes(extname(file.path)) &&
      config.filter(file)
  });

  const promises = files.map<Promise<ModuleEntry<T>>>(
    async ({ path, name }) => ({
      module: await import(path),
      path,
      name
    })
  );

  const modules = await Promise.all(promises);

  return modules;
}
