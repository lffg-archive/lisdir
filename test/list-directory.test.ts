import { join } from 'path';
import { listDirectory, Entry } from '../src/lisdir';

const MOCK_PATH = join(__dirname, 'mock', 'main');

const getData = (entries: Entry[]) => entries.map((entry) => entry.name);

describe('list directory', () => {
  it('should list the directory contents', async () => {
    const names = getData(await listDirectory(MOCK_PATH));
    expect(names).toMatchSnapshot();
  });

  it('should use the filter function', async () => {
    const names = getData(
      await listDirectory(MOCK_PATH, {
        filter: (file) => file.isFile()
      })
    );

    expect(names).toMatchSnapshot();
  });

  it('should be able to list the directory with recursion', async () => {
    const names = getData(await listDirectory(MOCK_PATH, { recursive: true }));
    expect(names).toMatchSnapshot();
  });
});
