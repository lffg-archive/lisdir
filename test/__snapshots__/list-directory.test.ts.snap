// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list directory should be able to list the directory with recursion 1`] = `
Array [
  "a.txt",
  "b.ts",
  "c.ts",
  "d.txt",
  "e.ts",
  "f.ts",
]
`;

exports[`list directory should list the directory contents 1`] = `
Array [
  "a.txt",
  "b.ts",
  "c.ts",
  "sub",
]
`;

exports[`list directory should use the filter function 1`] = `
Array [
  "a.txt",
  "b.ts",
  "c.ts",
]
`;
